07-folds-monoids/JoinList.hs

```diff
@@ -0,0 +1,244 @@
+We have provided the Sized module that defines the Size type, which is simply a newtype wrapper around an Int. In order to make Sizes more accessible, we have also defined the Sized type class which provides a method for obtaining a Size from a value.
+Use the Sized type class to write the following functions.
+-}
+--Exercise 2
+{-
+Implement the function
+indexJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> Maybe a
+indexJ finds the JoinList element at the specified index. If the index is out of bounds, the function returns Nothing. By an index in a JoinList we mean the index in the list that it represents. That is, consider a safe list indexing function
+(!!?) :: [a] -> Int -> Maybe a [] !!? _ = Nothing
+!!? i | i < 0 = Nothing (x:xs) !!? 0 =Justx (x:xs) !!? i = xs !!? (i-1)
+which returns Just the ith element in a list (starting at zero) if such an element exists, or Nothing otherwise. We also consider an updated function for converting join-lists into lists, just like jlbToList but ignoring the monoidal annotations:
+jlToList :: JoinList m a -> [a]
+jlToList Empty = []
+jlToList (Single _ a) = [a]
+jlToList (Append _ l1 l2) = jlToList l1 ++ jlToList l2
+We can now specify the desired behavior of indexJ. For any index i and join-list jl, it should be the case that
+   (indexJ i jl) == (jlToList jl !!? i)
+That is, calling indexJ on a join-list is the same as first convert- ing the join-list to a list and then indexing into the list. The point, of course, is that indexJ can be more efficient (O(log n) versus O(n), assuming a balanced join-list), because it gets to use the size annotations to throw away whole parts of the tree at once, whereas the list indexing operation has to walk over every element.
+-}
+indexJ :: (Sized b, Monoid b) => Int -> JoinList b a -> Maybe a
+indexJ n (Single _ a)
+    | n == 0    = Just a
+    | otherwise = Nothing
+indexJ n (Append m jl1 jl2)    
+    | n < jl1Size = indexJ n             jl1
+    | otherwise   = indexJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+indexJ _ _ = Nothing
+
+
+
+
+{-
+Implement the function
+dropJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> JoinList b a
+The dropJ function drops the first n elements from a JoinList. This is analogous to the standard drop function on lists. Formally, dropJ should behave in such a way that
+jlToList (dropJ n jl) == drop n (jlToList jl)
+-}
+dropJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+dropJ n (Single m a)
+    | n >= 1    = Empty
+    | otherwise = Single m a
+dropJ n (Append _ jl1 jl2)
+    | n < jl1Size = dropJ n             jl1 +++ jl2
+    | otherwise   = dropJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+dropJ _ _ = Empty
```

Comment:
We observe some redundancy, where there are 2 cases which map to `Empty`. To fix this we first understand the behaviour of guard clause:

```hs
f n | n > 5 = 1
f n | n < 1 = 2
f _ = 3
```


Running in ghci:
```
Prelude> f 6
1
Prelude> f 0
2
Prelude> f 3
3
Prelude> f 4
3
```

Whatever the guard clause doesn't catch is caught by the `f _` wildcard matching.

With this in mind, we can do the following:

```suggestion
dropJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
dropJ n (Single m a) | n < 1 = Single m a
dropJ n (Append _ jl1 jl2)
    | n < jl1Size = dropJ n             jl1 +++ jl2
    | otherwise   = dropJ (n - jl1Size) jl2
      where
        jl1Size :: Int
        jl1Size = getSize . size . tag $ jl1
dropJ _ _ = Empty
```

Reply:
@kwannoel Thanks! Could I check if inline guard clauses are syntactically okay?

```haskell
dropJ n (Single m a) | n < 1 = Single m a
```
vs
```haskell
dropJ n (Single m a) 
    | n < 1 = Single m a
```

Reply:
yes